### PR TITLE
BUG: tf.function cannot handle exception raised by custom op

### LINKED ISSUE #260: Raise exceptions based on tensor's properties
The build just failed yesterday due to some shape checking fragments in C++ code.
https://source.cloud.google.com/results/invocations/f48f6f28-c9a4-4912-b8c5-336f17167183

Currently, there are three approaches that can check tensor's properties and raise exceptions to block the following computation in addons (migrating from core TF):

1. Pure `if-else` and `raise` statement
https://github.com/tensorflow/addons/blob/d46dba1ef691f607a69e379fb3c0c2c16daec2fb/tensorflow_addons/image/dense_image_warp.py#L51-L53

2. `tf.debugging.*`
https://github.com/tensorflow/addons/blob/d46dba1ef691f607a69e379fb3c0c2c16daec2fb/tensorflow_addons/image/dense_image_warp.py#L70-L71

3. `OP_REQUIRES` in C++
https://github.com/tensorflow/addons/blob/d46dba1ef691f607a69e379fb3c0c2c16daec2fb/tensorflow_addons/custom_ops/image/cc/kernels/adjust_hsv_in_yiq_op.cc#L57-L59

However, none of them are robust enough in `tf.function` and in test cases with `@run_in_graph_and_eager_modes`. (https://github.com/tensorflow/addons/issues/138, https://github.com/tensorflow/addons/pull/257)

____

When checking core TensorFlow, I find this commit on Apr 10:
https://github.com/tensorflow/tensorflow/commit/4b4a39e7120b1c7744f9686bc6cce9363846d7e6#diff-68b5e47db1d9389c8d12852996845819

> Op raising `InvalidArgumentError` unless `x` is all negative. This can be
      used with `tf.control_dependencies` inside of `tf.function`s to block
      followup computation until the check has executed.

According to the doc, does it encourage us to use `tf.control_dependencies` inside `tf.function` to do something like shape checking?

### PR DESCRIPTION
**Pull request by facaiy**
Fix https://github.com/tensorflow/addons/issues/260

Hi, would you mind telling me where to add unit test?

### CODE REVIEW

**Comment by mdanatg** [comment]
Thank you for the fix!

I just realized that the integration test we used for this piece is not open sourced yet. So for now we should add a unit test in this same directory, similar to how other files do it - manually building an ErrorMetadataBase object with a suitable fake translated_stack should do the trick.

@bananabowl we should pull this into the 1.14 RC as well.

**Review by mdanatg — DISMISSED** [review]

**Comment by facaiy** [comment]
@mdanatg Thank you, Dan. I have added an unit test for the fix.

File: tensorflow/python/autograph/pyct/errors_test.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import textwrap
+
+from tensorflow.python.autograph.pyct import errors
+from tensorflow.python.platform import test
+
+
+class ErrorMetadataBaseTest(test.TestCase):
+
+  def test_get_message_when_frame_info_is_none(self):
```
Comment (mdanatg) [inline tensorflow/python/autograph/pyct/errors_test.py]: Nit: the test case is for when the code is None, rather than frame_info

File: tensorflow/python/autograph/pyct/errors_test.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    callsite_tb = [
+        ('<string>',
+         96,
+         'adjust_hsv_in_yiq',
+         None),
+        ('/usr/local/python/framework/ops.py',
+         1874,
+         '_create_c_op',
+         'raise ValueError(str(e))')]
+    cause_message = "ValueError: Shape must be at least rank 3."
```
Comment (mdanatg) [inline tensorflow/python/autograph/pyct/errors_test.py]: Consider using strings that are more evidently fakes: 'fake_function_name_1', 'test message', etc.

Comment (mdanatg) [inline tensorflow/python/autograph/pyct/errors_test.py]: Nit: please use single quotes for single-line strings, for consistency.

File: tensorflow/python/autograph/pyct/errors_test.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    em = errors.ErrorMetadataBase(
+        callsite_tb=callsite_tb,
+        cause_metadata=None,
+        cause_message=cause_message,
+        source_map={})
+    expected = """
```
Comment (mdanatg) [inline tensorflow/python/autograph/pyct/errors_test.py]: I worry that an exact string comparison is a change detector, it can break for unrelated changes. Could you use an assertRegex instead? Something like assertRegex(em.get_message(), 'adjust_hsv_in_yiq.*_create_c_op.*text message') should be sufficient.

**Review by mdanatg — DISMISSED** [review]
A few suggestions for style consistency.

**Comment by mdanatg** [comment]
Awesome, thank you! I left a few style comments, encouraged but optional.

**Comment by mdanatg** [comment]
@bananabowl this can be pulled into 1.14 as soon as tests are green.

**Comment by facaiy** [comment]
Thank you, I think all comments have been solved.

**Review by mdanatg — APPROVED** [review]
